packages/grid/test/keyboard-navigation.test.js

```diff
@@ -1934,4 +1938,162 @@ describe('keyboard navigation on column groups', () => {
+
+        const tabbableHeaderCell = getTabbableCells(header)[0];
+        expect(tabbableHeaderCell).not.to.be.ok;
+      });
+
+      it('should update tabbable header cell on header row unhide', async () => {
+        // Hide all header rows
+        mainGroup.headerRenderer = null;
+        subGroup.headerRenderer = null;
+        column.headerRenderer = null;
+        await nextFrame();
+
+        column.header = 'column';
+        await nextFrame();
+
+        const tabbableHeaderCell = getTabbableCells(header)[0];
+        expect(tabbableHeaderCell.offsetHeight).not.to.equal(0);
+      });
+    });
+
+    describe('updating tabbable cells - body', () => {
+      let body;
+
+      beforeEach(() => {
+        body = grid.$.items;
+      });
+
+      it('should update tabbable body cell on body row hide', async () => {
+        // Focus the second body row / make it tabbable
+        tabToBody();
+        down();
+
+        // Hide the second body row
+        grid.items = ['foo'];
+
+        await nextFrame();
+
+        // Expect the tabbable body cell to be on the first row
+        const tabbableBodyCell = getTabbableCells(body)[0];
+        expect(tabbableBodyCell.parentElement.index).to.equal(0);
+        expect(tabbableBodyCell.offsetHeight).not.to.equal(0);
+      });
+
+      it('should have no tabbable body cell when there are no rows', async () => {
+        // Remove all body rows
+        grid.items = [];
+
+        const tabbableBodyCell = getTabbableCells(body)[0];
+        expect(tabbableBodyCell).not.to.be.ok;
+      });
+
+      it('should update tabbable body cell on body row unhide', async () => {
+        // Remove all body rows
+        grid.items = [];
+        await nextFrame();
+
+        grid.items = ['foo', 'bar'];
+        await nextFrame();
+
+        const tabbableBodyCell = getTabbableCells(body)[0];
+        expect(tabbableBodyCell.parentElement.index).to.equal(0);
+        expect(tabbableBodyCell.offsetHeight).not.to.equal(0);
+      });
+
+      it('should remain on the first column after a tabbable cell row is hidden', async () => {
+        // Add a second column
+        const column = document.createElement('vaadin-grid-column');
+        grid.appendChild(column);
+        await nextFrame();
+
+        // Focus the second cell on the second row
+        tabToBody();
+        right();
+        down();
+
+        // Hide the second body row
+        grid.items = ['foo'];
+
+        // Focus the first row (should focus the first cell)
+        tabToBody();
+        // Hit down
+        down();
+
+        // Expect the focus to be on the first column
+        expect(getFocusedCellIndex()).to.equal(0);
+      });
+
+      it('should tab to body after reducing rows', async () => {
+        // Focus a cell on the second row
+        tabToBody();
+        down();
+
+        // Hide the second body row
+        grid.items = ['foo'];
+
+        // Navigate from header to body with tab
+        tabToHeader();
```

Comment:
Tried to reproduce this case on a page with a grid. What is the way for the user to get to the header, not resetting `_focusedItemIndex`?

```html
<body>
  <script type="module">
    import '@vaadin/grid/all-imports';

    const grid = document.querySelector('vaadin-grid');

    grid.items = [{ name: 'foo' }, { name: 'bar' }];

    setTimeout(() => {
      grid.items = [{ name: 'bar' }];
    },5000)
  </script>

  <vaadin-grid item-id-path="name">
    <vaadin-grid-selection-column auto-select frozen></vaadin-grid-selection-column>
    <vaadin-grid-tree-column frozen path="name" width="200px" flex-shrink="0"></vaadin-grid-tree-column>
    <vaadin-grid-column path="name" width="200px" flex-shrink="0"></vaadin-grid-column>
    <vaadin-grid-column path="name" width="200px" flex-shrink="0"></vaadin-grid-column>
    <vaadin-grid-column path="name" width="200px" flex-shrink="0"></vaadin-grid-column>
  </vaadin-grid>
</body>
```

Reply:
The issue does reproduce with the page you provided (if you revert the linked change):

https://user-images.githubusercontent.com/1222264/141291570-6dc788c9-bee4-4c99-943b-515a2f5cc3df.mp4

Reply:
Now I see, one more question, is it supposed that on the first Tab, after removing the last row, the first **body** row gets selected rather than the first **header** row? Looks suspicious because the test tabs to the header first.

https://user-images.githubusercontent.com/5039436/141292467-ea10d8a5-c97c-47be-9d71-830fabc34eb9.mov

Reply:
If the element you have focused gets removed from the DOM, the `<body>` becomes the new active element. Chrome probably keeps some anchor around where the focus was so the next tab press doesn't start again from the body but at the position where you were before. Seems that different browsers behave a bit differently in this scenario.

We could consider automatically focusing the new focusable when this happens, but not in the scope of this PR.

Reply:
Thanks for the explanation, now it is clear.